Chap_API_Reserved_Keys.tex

```diff
@@ -9,14 +9,16 @@ \chapter{Reserved Keys}
 \code{"pmix"}. By definition, reserved keys are provided by the host
 environment and the \ac{PMIx} server, and are required to be available at client
 start of execution. \ac{PMIx} clients and tools are therefore prohibited from
-posting reserved keys using the \refapi{PMIx_Put} \ac{API}.
+posting reserved keys.
 
-\ac{PMIx} implementations may choose to define their own
-custom-prefixed keys which may adhere to either the \emph{reserved} or the \emph{non-reserved} retrieval rules at the discretion of the implementation. Implementations may choose to provide such custom keys at client start of execution, but this is not required.
+%% TODO: Should we note that although values are required to be available at
+%% client start of execution, that some values can change during execution?
 
-Host environments may also opt to define their own custom keys. However, \ac{PMIx} implementations are unlikely to recognize such host-defined keys and will therefore treat them according to the \emph{non-reserved} rules described in Chapter \ref{chap:nrkeys}. Users
+Host environments may opt to define non-standardized reserved keys. 
```

Comment:
Are you saying a host environment is allowed to define a key that violates the "pmix" prefix rule - i.e., it starts with "pmix"?? This seems like a risky decision as there is no way to guarantee it won't conflict with something in an implementation. I would strongly advise against it.

Reply:
We thought that we were simplifying the existing text.  It was our understanding that "custom keys" was referring to "non-standardized reserved keys".   We took it to mean that because this whole selection is talking about reserved keys.    I guess you are saying that the old text assumes that jsm_foo or srun_bar are custom key which are not reserved (the definition of reserved is clear... starts with "pmix_") but they are made available at the start of execution and cannot be set or changed by the client?  We assumed that custom keys would be something like "pmix_jsm_foo" or "pmix_srun_bar", though I can see now that the text about "for a list of any custom prefixes" gave a hint to what was being said that we missed.  

Assuming my above explanation is accurate, then we have a few choices... 1) We can revert back to the old text 2) We can explain the old text better to avoid the confusion we experienced in thinking a custom-key was a non-standardized reserved key 3) We can move in the direction of the simplified explanation we gave possibly adding some a warning that using non-standarized reserved keys is dangerous because it may conflict with future standardization.

Reply:
Yeah, it probably needs clarification. "Custom keys" cannot start with "pmix". The point I was trying to make is that PMIx itself won't understand them, so you cannot use them to direct PMIx behavior. Pretty much like any non-reserved key, in hindsight, so maybe just saying that "custom keys must be non-reserved" (in better wording) would suffice.

Reply:
I'll just add that, the new text has the problem that non-standardized keys beginning with pmix_ can be used by implementations and 1) that could be confusing for user's who would expect them to be able to look them up in the standard 2) they might get defined by the standard later to behave differently than the implementation originally used them.   
The old text has the problem that 1) It attempts to define this class of keys that are neither reserved nor non-reserved.   They don't follow reserved retrieval rules, but they shouldn't be set by the user either.  It's really a not well defined set of keys 2)  User's have to constantly be on the look for conflicts with implementations.  Their old code may not run on a new version of the host environment because the host environment added a custom key that conflicted with the client's use of a non-reserved key.

Reply:
Perhaps the best way to describe this is:

- host environments, just like applications, may define their own non-reserved keys
- such keys must obey the non-reserved key rules regarding prefixing, and will be retrieved based on the rules for non-reserved keys
- host environments and applications are strongly advised to use appropriate prefixing on any non-reserved keys they define. You can even provide some guidance - e.g., it wouldn't be very smart for LSF to define keys starting with "ompi". It would, however, make sense for them to define keys starting with "lsf".

Users don't have to worry about conflicts with implementations as any PMIx implementation should be sticking with the reserved keys. You do indeed have to worry about conflicts with the host environment - but that is why users should be sensibly prefixing their keys. For example, you might expect an app named "foo" to prefix any key they define with "foo". Or to do like MPICH does and stick process IDs at the beginning.

As for an app not defining a value for a host environment key - I'm not sure we can/should say anything about it. Some environments might well define a key for apps to use when, for example, communicating a request to the RM. I don't see why PMIx cares either way. If the environment prohibits it, then they (a) should say so and (b) protect themselves against the inevitable violation of that prohibition.